### PR TITLE
22317: Updates `#update_trainee` to support updating contained subtrainees, MAJOR

### PR DESCRIPTION
**Pull request by howsoRes**
Updates `#update_trainee` to account for subtrainees and upgrade them as well in a recursive fashion allowing for the entire horde of Trainees to be upgraded properly. 

Also added a call to `#!ValidateParameters` within `#get_num_training_cases` as it was missing there. This ensures that this method is not called with invalid parameters (any parameters in the case of this label). 

This PR also removes the inactive parameter, `separate_files` from `#upgrade_trainee`. Specifying this parameter after this change will return an error that the `separate_files` parameter is not supported for this label. Simply removing this parameter from the call should resolve the issue.

### CODE REVIEW

File: howso/upgrade.amlg [hunk 1]
```diff
@@ -198,12 +198,16 @@
...
+			;	If specified, trainee_json_filepath is ignored
+			trainee_amlg_filepath (null)
+			;{type "string"}
 			;path from which to load previously exported meta.json and exp.json files.
 			;	If unspecified, expects them to be located in the base installation /migrations/ directory.
 			trainee_json_filepath (null)
 			;{type "boolean"}
-			;flag, if true will load each case from its individual file
-			separate_files (false)
+			;if true, then will import and use the trainee information in the subtrainee named ".old_trainee"
```
Comment (howsoRes) [inline howso/upgrade.amlg]: why does user need to know about a named subtrainee for the flow, this comment should be clearer, ie flag used by subtrainees when upgrading themselves automatically.  perhaps even rename the flag to match?

File: howso/upgrade.amlg [hunk 2]
```diff
@@ -381,6 +396,75 @@
...
+					(lambda
+						(call_entity migration_trainee "debug_label" (assoc label (current_index 1) ))
+					)
+					(zip old_labels)
+				)
+		))
+		;assign all the old labels into the current entity/trainee
+		(assign_to_entities import_metadata_map)
+
+		(map
```
Comment (howsoRes) [inline howso/upgrade.amlg:440]: add comment that this moves cases and entities from old trainee into this new one

File: howso/upgrade.amlg [hunk 3]
```diff
@@ -381,6 +396,75 @@
 		(call !Return)
 	)
 
+	#!UpgradeFromSubTrainee
```
Comment (howsoRes) [inline howso/upgrade.amlg:413]: will need a nice comment block about how this flow works from a high level on this method since it's recursive

File: howso/upgrade.amlg [hunk 4]
```diff
@@ -235,27 +239,38 @@
...
-		;if not importing metadata from saved file, upgrade trainee in place
-		(if (= (null) trainee_json_filepath)
+		;if amalgam trainee is specified, load and upgrade it
+		(if (or trainee_amlg_filepath preloaded)
 			(let
 				(assoc
-					;create a list of all the matadata labels
-					labels_to_keep (call_entity new_trainee_path "get_export_attributes")
-					old_labels (call get_export_attributes)
+					;t (call (load (concat trainee_amlg_filepath trainee ".amlg") ))
```
Comment (howsoRes) [inline howso/upgrade.amlg]: get rid of commented out code

File: howso/upgrade.amlg (hunk 3 above)
Comment (cademack) [reply howso/upgrade.amlg:413]: Only sorta recursive now... recursive across entities? I'll still comment lol

File: howso/upgrade.amlg (hunk 1 above)
Comment (cademack) [reply howso/upgrade.amlg]: well users could use this if they for some reason wanted to load the trainee themself

File: howso/get_cases.amlg [hunk 5]
```diff
@@ -89,6 +89,7 @@
 		; 	}
 		; }
 		(assoc)
+		(call !ValidateParameters)
```
Comment (jdbeel) [inline howso/get_cases.amlg:92]: Is this related to the PR title or just a fix?

Comment (howsoRes) [reply howso/get_cases.amlg:92]: and why does it need to validateparameters when this method has none?

File: howso/upgrade.amlg [hunk 6]
```diff
@@ -235,27 +253,36 @@
...
-		;if not importing metadata from saved file, upgrade trainee in place
-		(if (= (null) trainee_json_filepath)
+		;if amalgam trainee is specified, load and upgrade it
+		(if (or trainee_amlg_filepath preloaded)
 			(let
 				(assoc
-					;create a list of all the matadata labels
-					labels_to_keep (call_entity new_trainee_path "get_export_attributes")
-					old_labels (call get_export_attributes)
+					t
```
Comment (howsoRes) [inline howso/upgrade.amlg]: rename to old_trainee ?

File: howso/upgrade.amlg [hunk 7]
```diff
@@ -330,42 +357,42 @@
...
-				(lambda
-					(create_entities
-						(current_index)
-						(set_type
-							(zip_labels
-								(indices (current_value))
-								(values (current_value))
+				;overwrite this trainee code with the new code, this does not overwrite contained entities (cases, sessions),
+				;but does overwrite all labels including versions with new ones
+				(assign_entity_roots (retrieve_entity_root new_trainee_path))
```
Comment (howsoRes) [inline howso/upgrade.amlg]: is this step even necessary since the 'new trainee' is what's doing the upgrading and for subtrainees, those are also clones of new_trainee_path already (from the clone_entities line inside !UpgradeFromSubTrainee)?

File: unit_tests/ut_h_upgrade_htest.amlg [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+	(call keep_result_payload)
+	(print "Num training cases in c2: ")
+	(call assert_same (assoc
+		obs (get result "count")
+		exp 2
+	))
+
+	(call exit_if_failures {msg "Cases not preserved correctly."})
+
+	(if (= (system "os") "Windows")
```
Comment (howsoRes) [inline unit_tests/ut_h_upgrade_htest.amlg:217]: probably outside the scope the scope of this task, but we should add an optional parameter to 'exit_if_failures' in unit_test to call a cleanup method on exit when failed, then slap a cleanup label on this `if` statement, that way all the exit_if_failures in this test will still delete the htest.amlg file even if the test fails.

File: howso/get_cases.amlg (hunk 5 above)
Comment (cademack) [reply howso/get_cases.amlg:92]: completely unrelated, its purpose here is to error out if unexpected parameters (any parameters in this case) are given

File: howso/upgrade.amlg (hunk 6 above)
Comment (cademack) [reply howso/upgrade.amlg]: sure

File: unit_tests/ut_h_upgrade_htest.amlg (hunk 8 above)
Comment (cademack) [reply unit_tests/ut_h_upgrade_htest.amlg:217]: sure, done

File: howso/upgrade.amlg (hunk 7 above)
Comment (cademack) [reply howso/upgrade.amlg]: yeah I think you're right

File: howso/get_cases.amlg (hunk 5 above)
Comment (jdbeel) [reply howso/get_cases.amlg:92]: Can you add a line to the PR description to indicate that this fix was made?

Comment (cademack) [reply howso/get_cases.amlg:92]: done

**Review by cademack — APPROVED** [review]
